File: 1-chapter/Part_2/exponentWithPowerRule.py
def exponentWithPower(base, exponent):
    opStack = []
    while exponent > 1:
        if exponent % 2 == 0:
            opStack.append('square')
            exponent = exponent // 2
        elif exponent % 2 == 1:
            exponent -= 1
            opStack.append('multiply')
    result = base
    while opStack:
        op = opStack.pop()

        if op == 'multiply':
            result *= base
        elif op == 'square':
            result *= result

    return result


print(exponentWithPower(3, 6))
print(exponentWithPower(10, 3))
print(exponentWithPower(17, 10))

File: 1-chapter/Part_2/exponentWithPowerRule.js
function exponentWithPowerRule(a, n) {
    //Part1
    let opStake = [];
    while (n > 1) {
        if (n % 2 === 0) {
            opStake.push('square')
            n = Math.floor(n / 2)
        } else if (n % 2 === 1) {
            n -= 1;
            opStake.push('multiple')
        }
    }
    console.log(opStake)
    //Part2
    let result = a;
    while (opStake.length > 0) {
        let op = opStake.pop();

        if (op === 'multiple') {
            result = result * a
        } else if (op === 'square') {
            result = result * result
        }
    }

    return result
}

console.log(exponentWithPowerRule(3, 6))
console.log(exponentWithPowerRule(10, 3))
console.log(exponentWithPowerRule(17, 10))